app/models/user.rb

```diff
@@ -208,44 +208,57 @@ def self.upload_user_list(user_class, user_list, encoding)
     imported = nil
     parsed[:invalid_records] = ''
     User.transaction do
-      imported = user_class.import user_columns, users, on_duplicate_key_update: {
-        conflict_target: [:user_name],
-        columns: [:last_name, :first_name, :section_id, :email, :id_number, :display_name, :time_zone]
-      }
-      User.where(id: imported.ids).each do |user|
-        if user_class == Ta
-          # This will only trigger before_create callback in ta model, not after_create callback
-          user.run_callbacks(:create) { false }
+      users.each do |user|
+        user.length.times do |i|
+          if user[i].nil?
+            user[i] = '-'
+          end
         end
-        user.validate!
-      rescue ActiveRecord::RecordInvalid
-        error_message = user.errors
-                            .messages
-                            .map { |k, v| "#{k} #{v.flatten.join ','}" }.flatten.join MarkusCsv::INVALID_LINE_SEP
-        parsed[:invalid_records] += "#{user.user_name}: #{error_message}"
       end
-      unless parsed[:invalid_records].empty?
-        raise ActiveRecord::Rollback
+
+      user_hash = users.collect { |record| Hash[user_columns.zip record] }
+      imported = user_class.upsert_all(user_hash, unique_by: :user_name) unless user_columns.empty? ||
+        users.empty? || :user_name.nil? || user_hash.empty?
+
+      unless imported.nil?
+        User.where(id: imported.rows.flatten).each do |user|
```

Comment:
You use `imported.rows.flatten` a few different times, so this would be good to save in a variable.